driver-mapping/src/main/java/com/datastax/driver/mapping/Mapper.java

```diff
@@ -72,26 +72,19 @@ public T apply(ResultSet rs) {
                 return Mapper.this.map(rs);
             }
         };
+        this.preparedQueries.clear();
     }
 
     Session session() {
         return manager.getSession();
     }
 
-    PreparedStatement getPreparedQuery(QueryType type) {
-        PreparedStatement stmt = preparedQueries.get(type);
-        if (stmt == null) {
-            synchronized (preparedQueries) {
-                stmt = preparedQueries.get(type);
-                if (stmt == null) {
-                    String query = type.makePreparedQueryString(tableMetadata, mapper);
-                    logger.debug("Preparing query {}", query);
-                    stmt = session().prepare(query);
-                    Map<QueryType, PreparedStatement> newQueries = new HashMap<QueryType, PreparedStatement>(preparedQueries);
-                    newQueries.put(type, stmt);
-                    preparedQueries = newQueries;
-                }
-            }
+    PreparedStatement getPreparedQuery(QueryType type){
+        String queryString = type.makePreparedQueryString(tableMetadata, mapper, saveOptions);
+        PreparedStatement stmt = preparedQueries.get(queryString);
+        if (stmt == null){
+            stmt = session().prepare(queryString);
+            preparedQueries.putIfAbsent(queryString, stmt);
```

Comment:
This is not the same as the previous version, which guaranteed that, if two threads try to prepare the same query at the same time:
1. the query will be prepared only once
2. the two threads will get the same object (actually a moot point because the underlying `session` already guarantees it)

Point 1 exposes us to getting the warning "Re-preparing already prepared query {}...", which is not great.